docs/nodemcu-pil.md

```diff
@@ -0,0 +1,40 @@
+-  ROTables are typically declared at the end of the module to minimise the need for forward references and use the `LROT` macros described in the NRM.  Note that ROTables only support static declaration of string keys and the value types: C function, Lightweight userdata, Numeric, ROTable.  ROTables can also have ROTable metatables.
+   -  Whilst the ROTable search algorithm is a simply linear scan of the ROTable entries, the runtime also maintains a LRU cache of ROTable accesses, so typically over 95% of ROTable accesses bypass the linear scan and do a direct access to the appropriate entry.
+   -  ROTables are also reasonable lightweight and well integrated into the Lua runtime, so the normal metamethod processing works well.  This means that developers can use the `__index` method to implement other key and value typed entries through an index function.
+-  NodeMCU modules are intended to be compilable against both our Lua 5.1 and Lua 5.3 runtimes.  The NRM discusses the implications and constraints here.  However note that:
+   -  We have back-ported many new Lua 5.3 features into the NodeMCU Lua 5.1 API, so in general you can use the 5.3 API to code your modules.  Again the NRM notes the exceptions where you will either need variant code or to decide to limit yourself to the the 5.3 runtime. In this last case the simplest approach is to `#if LUA_VERSION_NUM != 503` to disable the 5.3 content so that 5.1 build can compile and link.   Note that all modules currently in the `app/modules` folder will compile against and execute within both the Lua 5.1 and the 5.3 environments.
+   -  Lua 5.3 uses a 32-bit representation for all numerics with separate subtypes for integer (stored as a 32 bit signed integer) and float (stored as 32bit single precision float).  This achieves the same RAM storage density as Lua 5.1 integer builds without the loss of use of floating point when convenient.  We have therefore decided that there is no benefit in having a separate Integer 5.3 build variant.
+ -  We recommend that developers make use of the full set of `luaL_` API calls to minimise code verbosity.  We have also added a couple of registry access optimisations that both simply and improve runtime performance when using the Lua registry for callback support.
+    - `luaL_reref()` replaces an existing registry reference in place (or creates a new one if needed).  Less code and faster execution than a `luaL_unref()` plus `luaL_ref()` construct.
+    - `luaL_unref2()` does the unref and set the static int hook to `LUA_NOREF`.
+
+Rather than include simple examples of module templates, we suggest that you review the modules in our GitHub repository, such as the [`utf8`](https://github.com/nodemcu/nodemcu-firmware/tree/release/app/lua53/lutf8lib.c) library.  Note that whilst all of the existing modules in `app/modules` folder compile and work, we plan to do a clean up of the core modules to ensure that they conform to best practice.
```

Comment:
> we suggest that you review the modules in our GitHub repository, such as the `utf8`

Feels odd to me as that's not a "module" as such.

Reply:
We actually have two groups of modules:
-  the "standard" Lua ones: `bit`, `coroutine`, `debug`, `math`, `string`, `table`, `utf8`.  Even the base functions (such as `print`) are implemented as really of the module `ROM` which also includes all of the other linked modules as members.  `ROM` is linked to `_G` by its `__index` metamethod.  So `_G.ROM.print`, `_G.print`, `ROM.print`, `print` all resolve to the same print function.
- the NodeMCU ones linked in from `app/modules`

In terms of their code they are structured the same way, use the same Lua C APIs.  However, the difference was that the authors of the Lua ones really knew their way around the C API, which more than can be said for most of our module authors, who essentially copied Zeroday's coding style (written at a point when he was still leaning the C API).  Hence, in general the standard Lua module implementation are a lot tighter / neater than our modules. :scream: 

What I have done is on the earlier line "`Hence a NodeMCU C library module typically has a standard layout`" to add the explanatory qualifier "`that parallels that of the standard Lua library modules and uses the same C API to access the Lua runtime:`".

One of the thing on my / @nwf TODO list is to clean up our core modules. We can update this document linking to a cleaned up module once we've done at least one, so we don't fell ashamed when comparing them to the Lua ones. :rofl: